kernel/src/net/socket/util/options.rs

```diff
@@ -26,11 +32,102 @@ impl SocketOptionSet {
             sock_errors: None,
             reuse_addr: false,
             reuse_port: false,
-            send_buf: SEND_BUF_LEN as u32,
-            recv_buf: RECV_BUF_LEN as u32,
+            send_buf: TCP_SEND_BUF_LEN as u32,
+            recv_buf: TCP_RECV_BUF_LEN as u32,
             linger: LingerOption::default(),
         }
     }
+
+    /// Return the default socket level options for udp socket.
+    pub fn new_udp() -> Self {
+        Self {
+            sock_errors: None,
+            reuse_addr: false,
+            reuse_port: false,
+            send_buf: UDP_SEND_PAYLOAD_LEN as u32,
+            recv_buf: UDP_RECV_PAYLOAD_LEN as u32,
+            linger: LingerOption::default(),
+        }
+    }
+
+    /// Get option. Return Ok(true) if it need to clear the socket errors
+    pub fn get_option(&self, option: &mut dyn SocketOption) -> Result<bool> {
+        // Note that the socket error has to be handled separately, because it is automatically
+        // cleared after reading.
+        match_sock_option_mut!(option, {
+            socket_errors: SocketError => {
+                let sock_errors = self.sock_errors();
+                socket_errors.set(sock_errors);
+
+                return Ok(true);
```

Comment:
This will unnecessarily create races because it asks the caller to acquire the lock again and clear the error code.

How about something like:
```rust
impl SocketOptionSet {
    /// Gets and clears the socket error.
    ///
    /// When processing the `getsockopt` system call, the socket error is automatically cleared
    /// after reading. This method should be called to provide this behavior.
    fn get_and_clear_sock_errors(&mut self, option: &mut SocketError) {
        // ...
    }

    /// Gets socket-level options.
    ///
    /// Note that the socket error has to be handled separately, because it is automatically
    /// cleared after reading. This method does not handle it. Instead,
    /// [`Self::get_and_clear_socket_errors`] should be used.
    pub fn get_option(&self, option: &mut dyn SocketOption) -> Result<()> {
        // ..
    }
}
```

Then we can write:
```rust
       match_sock_option_mut!(option, {
            socket_errors: SocketError => {
                self.options.write().get_and_clear_sock_errors(&socket_errors);
                return Ok(());
            },
            _ => (),
        }

        let options = self.options.read();

        match options.socket.get_option(option) {
            // ...
        }

        // ... (TCP options)
```